core/src/main/scala/kafka/server/AddPartitionsToTxnManager.scala

```diff
@@ -0,0 +1,180 @@
+        new TransactionDataAndCallbacks(
+          new AddPartitionsToTxnTransactionCollection(1),
+          mutable.Map[String, AddPartitionsToTxnManager.AppendCallback]()))
+
+      val currentTransactionData = currentNodeAndTransactionData.transactionData.find(transactionData.transactionalId)
+
+      // Check if we already have txn ID -- if the epoch is bumped, return invalid producer epoch, otherwise, the client likely disconnected and 
+      // reconnected so return the retriable network exception.
+      if (currentTransactionData != null) {
+        val error = if (currentTransactionData.producerEpoch() < transactionData.producerEpoch())
+          Errors.INVALID_PRODUCER_EPOCH
+        else 
+          Errors.NETWORK_EXCEPTION
+        val topicPartitionsToError = mutable.Map[TopicPartition, Errors]()
+        currentTransactionData.topics().forEach { topic =>
+          topic.partitions().forEach { partition =>
+            topicPartitionsToError.put(new TopicPartition(topic.name(), partition), error)
+          }
+        }
+        val oldCallback = currentNodeAndTransactionData.callbacks(transactionData.transactionalId())
+        currentNodeAndTransactionData.transactionData.remove(transactionData)
+        oldCallback(topicPartitionsToError.toMap)
+      }
+      currentNodeAndTransactionData.transactionData.add(transactionData)
+      currentNodeAndTransactionData.callbacks.put(transactionData.transactionalId(), callback)
+      wakeup()
+    }
+  }
+
+  private class AddPartitionsToTxnHandler(node: Node, transactionDataAndCallbacks: TransactionDataAndCallbacks) extends RequestCompletionHandler {
+    override def onComplete(response: ClientResponse): Unit = {
+      // Note: Synchronization is not needed on inflightNodes since it is always accessed from this thread.
+      inflightNodes.remove(node)
+      if (response.authenticationException() != null) {
+        error(s"AddPartitionsToTxnRequest failed for broker ${config.brokerId} with an " +
+          "authentication exception.", response.authenticationException)
+        transactionDataAndCallbacks.callbacks.foreach { case (txnId, callback) =>
+          callback(buildErrorMap(txnId, Errors.forException(response.authenticationException()).code()))
+        }
+      } else if (response.versionMismatch != null) {
+        // We may see unsupported version exception if we try to send a verify only request to a broker that can't handle it. 
+        // In this case, skip verification.
+        warn(s"AddPartitionsToTxnRequest failed for broker ${config.brokerId} with invalid version exception. This suggests verification is not supported." +
+          s"Continuing handling the produce request.")
+        transactionDataAndCallbacks.callbacks.values.foreach(_(Map.empty))
+      } else if (response.wasDisconnected() || response.wasTimedOut()) {
+        warn(s"AddPartitionsToTxnRequest failed for broker ${config.brokerId} with a network exception.")
```

Comment:
It's probably used for to log the destination broker on which the request fails. Ditto in other logging.